backend/msgpack/writer.go

```diff
@@ -240,21 +272,54 @@ func (w *writer) encodeWithLock(
 		return nil
 	}
 
-	// Otherwise we get a new buffer from pool, copy the bytes exceeding the
-	// flush size to it, swap the new buffer with the old one, and flush out
-	// the old buffer.
-	newBufferedEncoder := w.encoderPool.Get()
+	// Otherwise we prepare to enqueue the current buffer.
+	oldBufferedEncoder := w.prepareEnqueue(encoder, sizeBefore)
+
+	encoder.Unlock()
+
+	// Write out the buffered data.
+	if err := w.queue.Enqueue(oldBufferedEncoder); err != nil {
+		return err
+	}
+	w.metrics.buffersEnqueued.Inc(1)
+
+	return nil
+}
+
+// prepareEnqueue gets a new buffer from pool, copies the bytes exceeding
+// the flush size to it, swaps the new buffer with the old one in the
+// encoder, and returns the old buffer.
+func (w *writer) prepareEnqueue(
+	encoder *lockedEncoder,
+	sizeBefore int,
+) msgpack.BufferedEncoder {
```

Comment:
Change the return type to `msgpack.Buffer` to prevent caller misusing the return value.